buildSrc/src/main/java/org/elasticsearch/gradle/ReaperPlugin.java

```diff
@@ -0,0 +1,141 @@
+
+            @Override
+            public void buildFinished(BuildResult result) {
+                if (reaperProcess != null) {
+                    try {
+                        reaperProcess.getOutputStream().close();
+                        project.getLogger().info("Waiting for reaper to exit normally");
+                        if (reaperProcess.waitFor() != 0) {
+                            throw new GradleException("Reaper process failed. Check log at "
+                                + baseDirectory.resolve("error.log") + " for details");
+                        }
+                    } catch (Exception e) {
+                        throw new RuntimeException(e);
+                    }
+
+                }
+            }
+        });
+    }
+
+    /**
+     * Register a pid that will be killed by the reaper.
+     */
+    public void registerPid(String serviceId, long pid) {
+        String[] killPidCommand = OS.<String[]>conditional()
+            .onWindows(() -> new String[]{"Taskill", "/F", "/PID", String.valueOf(pid)})
+            .onUnix(() -> new String[]{"kill", "-9", String.valueOf(pid)}).supply();
+        registerCommand(serviceId, killPidCommand);
+    }
+
+    /**
+     * Register a system command that will be run by the reaper.
+     */
+    public void registerCommand(String serviceId, String... command) {
+        ensureReaperStarted();
+
+        try {
+            Files.writeString(baseDirectory.resolve(serviceId), String.join(" ", command));
+        } catch (IOException e) {
+            throw new UncheckedIOException(e);
+        }
+    }
+
+    public void unregister(String serviceId) {
+        try {
+            Files.deleteIfExists(baseDirectory.resolve(serviceId));
+        } catch (IOException e) {
+            throw new UncheckedIOException(e);
+        }
+    }
+
+    private synchronized void ensureReaperStarted() {
+        if (reaperProcess == null) {
+            try {
+                // copy the reaper jar
+                Path jarPath = project.getBuildDir().toPath().resolve("reaper").resolve("reaper.jar");
+                Files.createDirectories(jarPath.getParent());
+                InputStream jarInput = ReaperPlugin.class.getResourceAsStream("/META-INF/reaper.jar");
+                try (OutputStream out = Files.newOutputStream(jarPath)) {
+                    jarInput.transferTo(out);
+                }
+
+                // ensure the input directory exists
+                Files.createDirectories(baseDirectory);
+
+                // start the reaper
+                ProcessBuilder builder = new ProcessBuilder(
+                    "java", "-cp", jarPath.toString(), "org.elasticsearch.gradle.reaper.Reaper", baseDirectory.toString());
+                File logFile = baseDirectory.resolve("reaper.log").toFile();
+                builder.redirectOutput(logFile);
+                builder.redirectError(logFile);
+                reaperProcess = builder.start();
```

Comment:
One missing bit in terms of error handling here is that we won't get any kind of exception if we simply fail to launch to reaper itself since `start()` is effectively async.

We should probably check for `isAlive()` at times. Perhaps whenever we call one of the `register()` methods, and also in `buildFinished()`. That way if for some reason the process shit the bed when we didn't expect we'd _eventually_ be notified.